### ISSUE TITLE
Improve Documentation To Highlight Reliance on Implementation Details of Assert As Bad

### ISSUE THREAD

**Issue by aolszowka** [nunit/docs#294]
This is a continuation of this discussion: #168 I completely agree with @CharliePoole on this, but I would suggest that  the documentation be improved to highlight the situation.

-- Begin Documentation Improvement Suggestion --

### Behavior of Asserts
```csharp
[Test]
public void TestAssertCatch()
{
    try
    {
        Assert.That(true, Is.False);
    }
    catch
    {
    }
}
```

When looking at this example you may expect NUnit to Assert that true != false. However in versions of NUnit Prior to 3.6.0.0 this test would pass due to an implementation detail of Assert. In newer versions of NUnit this will Assert (and fail the test) as a design decision of the NUnit.org Team.

In cases where this is used, it was probably used in scenarios where better APIs could be leveraged.

Consider the following test which is trying to Assert that an Exception is thrown/not thrown:

```csharp
public void TestException()
{
    try
    {
        GuessTheNumber(42);

        // This code only executes if no exception has occurred
        Assert.Fail("An expected exception did not occur");
    }
    catch(Exception ex)
    {
        Assert.That(ex.Message, Is.EqualTo("The Answer"), "An expected exception did not occur");
    }
}

public static void GuessTheNumber(int number)
{
    if(number == 42)
    {
        throw new Exception("The Answer");
    }
}
```

Instead this should be rewritten to use either Assert.That or Assert.Throws like so:

```csharp
[Test]
public void TestExceptionAssertThat()
{
    Assert.That(() => GuessTheNumber(42), Throws.InstanceOf<Exception>().With.Property("Message").EqualTo("The Answer"), "An expected exception did not occur");
}
```

Or Assert.Throws Style

```csharp
[Test]
public void TestExceptionAssertThrows()
{
    Exception ex = Assert.Throws<Exception>(() => GuessTheNumber(42), "An expected exception did not occur");
    Assert.That(ex.Message, Is.EqualTo("The Answer"), "An expected exception did not occur");
}
```

-- End Documentation Improvement --

As Charlie mentioned:

>I have asked the submitter of the issue to give a motivating example that would help us meet the need

It would be nice to hear what other scenarios end users are encountering that cause them to need to capture the exception of the Assert. The above example is just what we found in our code base during our 2.6.4->3.11.0 Convert.

Thank you for this tool; it is invaluable to us.

**Comment by SamuelEnglard** [comment]
In our code base we also have a need to check the members of the actual exception thrown. We've implemented a mock logger for testing that throws exceptions for error level logs and up. To check what was logged when testing we need to actually look at the exception.

**Comment by aolszowka** [comment]
@SamuelEnglard I am having trouble understanding the scenario you describe, could you share a small code snippet?

**Comment by SamuelEnglard** [comment]
@aolszowka this help?

```csharp
[Test]
public void LogError()
{
    Assert.That(() => AMethod(mockLoggerInstance), Throws.InstanceOf<MockLoggerException>().And.Property("LogLevel").EqualTo(LogLevel.Error);
}
```

**Comment by aolszowka** [comment]
Ah yup, so you guys are using the Assert.That Throws pattern. Thank you :)

**Comment by SamuelEnglard** [comment]
Yup, though I have helper class/methods to make that look nicer so for us it looks like

```csharp
[Test]
public void LogError()
{
    Assert.That(() => AMethod(mockLoggerInstance), Logs.Error());
}
```
